pint_xarray/conversion.py

```diff
@@ -122,18 +123,36 @@ def dataset_from_variables(variables, coords, attrs):
 def attach_units_dataset(obj, units):
     attached = {}
     rejected_vars = {}
+
+    indexed_variables = obj.xindexes.variables
     for name, var in obj.variables.items():
+        if name in indexed_variables:
+            continue
+
         unit = units.get(name)
         try:
             converted = attach_units_variable(var, unit)
             attached[name] = converted
         except ValueError as e:
             rejected_vars[name] = (unit, e)
 
+    ds_xindexes = obj.xindexes
+    new_indexes, new_index_vars = ds_xindexes.copy_indexes()
+
+    for idx, idx_vars in ds_xindexes.group_by_index():
+        idx_units = {name: units.get(name) for name in idx_vars.keys()}
+        new_idx = PintMetaIndex(index=idx, units=idx_units)
+        new_indexes.update({k: new_idx for k in idx_vars})
+        new_index_vars.update(new_idx.create_variables(idx_vars))
+
+    new_coords = Coordinates(new_index_vars, new_indexes)
```

Comment:
Could you try `Coordinates._construct_direct(new_index_vars, new_indexes)` to see if there still a warning due to safe_cast_to_index?

Reply:
that does remove the warning. What do you suggest we should do? Keep the `_construct_direct`?

Reply:
We are problably hitting https://github.com/pydata/xarray/pull/8107#discussion_r1311214263 here.

Although that's not very nice relying on a "private" method here, I think that it is probably best to keep `_construct_direct` until the issue is properly addressed in Xarray.

Reply:
I'm not in a hurry to cut a release with this feature included, so we can take our time figuring out how to address this in xarray, and once that's done I can update the code here.

Reply:
I don't think that `Coordinates._construct_direct` will go anywhere anytime soon, so it is reasonably safe using it for now. 

https://github.com/pydata/xarray/pull/8124 is tricky and might take a while, perhaps you might not want to wait for a proper fix before merging this and cut a new release.

Reply:
> perhaps you might not want to wait for a proper fix before merging this and cut a new release.

... especially since you'd like to present this as an example in the blog post, right?

Given that I'm both upstream *and* downstream here, I guess the risk is low enough that we can prioritize moving faster